Add GeneralizedHarmonic executable

prayush opened the pull request:
## Proposed changes

This PR adds the executable for the `GeneralizedHarmonic` system, which evolves 3D `AnalyticSolution` initial data. It also adds an observer to write out constraints and errors in evolved tensors (against the analytic solution). 

**CURRENT:**

- All dependencies merged

_______________________________________________
**EDIT (08/12/2019):**

- Once #1633 is merged, I'll replace the observer coded up here with generic observers. The output with generic observers (using compute tags from #1633) should remain identical to what it is now with the temporary observer. Please consider (`Observe.hpp`) as only serving to help characterize the actual evolution at the moment and review rest of the PR.

_______________________________________________

**EDIT (06/24/2019):**

This PR has been rebased onto PDALs (#1535), and now uses them. It **still** depends on:
- #1322 (first commit of this PR)

**Please review more recent commits than this only.**

_______________________________________________

**EDIT (06/06/2019):**

After rebase with current develop, this PR now depends only on:
- #1322 (first commit of this PR)

**Please review more recent commits than this only.**

_______________________________________________

**EDIT (05/22/2019):**

After rebase with current develop, this PR now depends only on:
- #1444 (first commit of this PR)
- #1499 (second commit of this PR)
- #1322 (third commit of this PR)

**Please review more recent commits than these only.**

_______________________________________________

All code for this PR is in the **most recent commit**. It depends on other PRs, whose contents have been committed as **6** previous commits. They are, in chronological order:

- 1st commit contains code from #1468 , #1469 , #1470 , #1471 , #1473 , #1474 , #1475 , #1476 , #1477  and #1478
- 2nd commit contains code from #1444 
- 3rd commit contains code from #1499 
- 4th commit contains code from #1497 
- 5th commit contains code from #1322 
- 6th commit contains code from #1496 .

Therefore: **please review the most recent commit only**.


### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

osheamonn commented on src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp:
```diff
@@ -0,0 +1,281 @@
+ * \f$\Phi_{iab}\f$ to disk as volume data.
+ * - The RMS error of \f$\Psi\f$ and \f$\Pi\f$ are written to disk.
+ */
+struct Observe {
```
Is it possible for this to use the generic observations infrastructure that was added a few weeks ago? The GrMhd executable currently uses it, and If it can be used for GH I think this file would then be unnecessary.

prayush replied:
This is a good point. The new generic observations infrastructure, that works through `EventsAndTriggers`, consists primarily of two observers: [dg::Events::ObserveErrorNorms](https://spectre-code.org/classdg_1_1events_1_1observeerrornorms_3_01volumedim_00_01tmpl_1_1list_3_01tensors_8_8_8_01_4_00_01eventregistrars_01_4) and [dg::Events::ObserveFields](https://spectre-code.org/classdg_1_1events_1_1observefields_3_01volumedim_00_01tmpl_1_1list_3_01tensors_8_8_8_01_4_00_01tc6487cc51554fb0bf44e7a56facd638b). It turns out that:

- `ObserveErrorNorms` can compute and write L2 norms for errors in user-provided variables against their *analytic* solution values. While GeneralizedHarmonic evolves `Psi, Phi, Pi`, the underlying KerrSchild solution does not provide any of these variables (it, in turn, only provides our usual 3+1 quantities). Therefore, we cannot use `ObserveErrorNorms` here to write out L2 norms of errors in `Psi, Phi, Pi`.
- `ObserveFields` can write out either (A) errors in tensors against their analytic solution counterparts, or (B) tensors directly. I can certainly use (B) here to write out constraint violations  in volume, *but* not their L2 norms. Unfortunately, however, I cannot use `ObserveFields` to write out *errors* in `Psi, Phi, Pi` in volume (again, because these vars arent available in KerrSchild). 

In `GeneralizedHarmonic::Actions::Observe` we are currently observing:
- Reduction: errors against analytic solution for `Psi, Phi, Pi`, L2 norms of gauge and other constraints;
- Volume: errors against analytic solution for `Psi, Phi, Pi`, L2 norms of gauge and other constraints.

Both `ObserveErrorNorms` and `ObserveFields` are inapplicable to observing errors in `Psi, Phi, Pi`, and this means that _at best_ I can offload some of the current volume observation to `ObserveFields` while keeping others in `GeneralizedHarmonic::Actions::Observe`. This last option has been discouraged though, as performing more than one volume observations at the same time can lead to unpredicted behavior (see warning [here](https://spectre-code.org/classdg_1_1events_1_1observeerrornorms_3_01volumedim_00_01tmpl_1_1list_3_01tensors_8_8_8_01_4_00_01eventregistrars_01_4)) **[1]**.

I think what we need is a generic `ObserveFieldNorms` reduction observer and a similar volume observer, both of which write out L2 norms of user-provided fields. These will be useful in writing out norms of constraint violations **[2]**. Then we need compute items taking the difference between evolved and analytic field values. Once we have *both*, we can switch to the new observation infrastructure completely, and abandon `GeneralizedHarmonic::Actions::Observe`. For now, I can't see how we can avoid keeping it though.

Thoughts?

**[1]** ..unless I am misreading the warning for `ObserveFields`.

**[2]** I have written up a quick example of what I mean as the last commit on [this dummy PR](https://github.com/prayush/spectre/pull/14). There I've added a generic `ObserveNorms` [reduction observer](https://github.com/prayush/spectre/blob/4bdc4675c2723c503c382fa34c7443bc4d70fde2/src/Evolution/DiscontinuousGalerkin/ObserveNorms.hpp) to write L2 norms of input tensors, and used it to observe constraint violations. If this is the direction we want to pursue, the next step would be to write compute items to get differences between evolved and analytic values of `Psi, Phi, Pi`, and then write a similar `ObserveFieldNorms` observer for outputting volume data.

kidder replied:
What is `ObserveFieldNorms` supposed to do?  Isn't `ObserveNorms` already enough?  
Also, I believe `ObserveFields` is already capable of (or easily extended to) observing any volume data

prayush replied:
@kidder  Currently:
- `ObserveNorms` **cannot** write out norms of any constraints. `ObserveFieldNorms` is supposed to be able to do that.
- `ObserveFields` can write out tensorial versions of constraints in volume, but **not** their norms in volume.

wthrowe replied:
Would writing a compute item that takes a pointwise norm be sufficient to allow the generic observers to be used instead of this?

Edit: Actually, isn't that what #1633 does?  Can you just wait for that to be merged?

geoffrey4444 replied:
Pretty sure I already have this on a branch...will check and update this pr if so

nilsvu replied:
With the new option tag handling written by @kidder it should also be possible to use cache tags in compute items (to be confirmed by @kidder, may require a small extension of his PR). Once that works, I have the compute item coded up to put analytic solutions of fields into the DataBox. That will be necessary to feed into the pointwise norm compute item that @wthrowe mentioned.

prayush replied:
@wthrowe @geoffrey4444 @nilsleiffischer - yes, my plan was to switch to generic observers once we have #1633 . Sorry, I should have made this clear in the description here once #1633 was put up. Until that PR is merged, please consider the GH observer here as placeholder (i.e. which can be used to check the actual evolution, but need not be reviewed).

prayush replied:
this file has been removed - replaced by generic observers

osheamonn commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,180 @@
+
+  static constexpr OptionString help{
+      "Evolve a generalized harmonic analytic solution.\n\n"
+      "The analytic solution is: Minkowski\n"
```
Minkowski -> KerrSchild

prayush replied:
fixed

prayush commented:
@geoffrey4444 - this is now rebased onto current develop. 

I had to make one change that I did not expect: `UnnormalizedFaceNormal` needed to inherit from both `db::ComputeTag` as well as `db::SimpleTag`, since once the prefix-tag `Tags::Normalized` is applied to it, [this bit of code](https://github.com/sxs-collaboration/spectre/blob/develop/src/DataStructures/DataBox/DataBoxTag.hpp#L521) doesn't like it that the wrapped tag is **not** a `SimpleTag`. Prefix tags are, apparently, only meant to wrap simple tags.. :smile: 

**Edit (PK):** The comment above is no longer relevant since `UnnormalizedFaceNormal` was split into a compute and a simple tag.

prayush commented:
@geoffrey4444 and @markscheel : the GeneralizedHarmonic executable added in this PR has now been rebased to use phase dependent action lists (PDALs) [c.f. #1535]

hrueter commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,262 @@
+      GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild>>;
+  using boundary_condition_tag = analytic_solution_tag;
+  using normal_dot_numerical_flux = OptionTags::NumericalFlux<
+      // dg::NumericalFluxes::LocalLaxFriedrichs<system>>;
```
Remove or comment why this is here.

prayush replied:
removed

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,195 @@
+
+    // set time derivatives of GaugeH = 0
+    const auto& dt_gauge_source =
+        make_with_value<tnsr::a<DataVector, Dim, Inertial>>(lapse, 0.);
```
I think this should be set using `GeneralizedHarmonic::Tags::SpacetimeDerivGaugeHCompute`.

geoffrey4444 replied:
This is just for setting the time derivatives. `SpacetimeDerivGaugeHCompute` is used to get the spacetime derivatives on line 213

hrueter replied:
For consistency the time derivative should be chosen according to the gauge that is used. Of course the result should be the same here, but this being a prototype for future GHG evolutions, I think we should implement the general case.

prayush replied:
There is a misunderstanding here:
 - the original comment here by @hrueter from 18days ago was for the line above in which `dt_gauge_source` was being explicitly initialized to `0`.
 - Now:
   - `dt_gauge_source = dt<GaugeH>` is provided through the compute tag `GeneralizedHarmonic::SpacetimeDerivDampedHarmonicHCompute`. @geoffrey4444 - `SpacetimeDerivGaugeHCompute` is NO longer used here.
   - Only `dt<Initial data's GaugeH>`, which is needed by the damped harmonic gauge, is being initialized to `0`. github has simply moved over this conversation to the line initializing `dt_initial_gauge_source`.

geoffrey4444 replied:
In real life, I think we'll only have two cases: i) analytic solutions, where we will always want the initial gauge to be stationary, and ii) numerical solutions, where we will read in the time derivatives. For now, I favor hard-coding zero here with a comment noting that this will have to be extended to handle numerical initial data or analytic initial data whose gauge is not initially stationary.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp:
```diff
@@ -0,0 +1,325 @@
+  };
+  struct ObserveAtT0 {
+    using type = bool;
+    static constexpr OptionString help = {"If true observe at t=0"};
```
Maybe write "If true **also** observe at t=0". I thought at first this means "**only** observe at t=0".

prayush replied:
this file has been removed - replaced by generic observers

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp:
```diff
@@ -0,0 +1,325 @@
+    if (time_id.slab_number() >= 0 and time_id.time().is_at_slab_start() and
+        static_cast<size_t>(time_id.slab_number()) %
+                Parallel::get<ObserveNSlabs>(cache) ==
+            0) {
```
Use brackets to make clear that `%` is evaluated before `==`.

prayush replied:
this file has been removed - replaced by generic observers

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp:
```diff
@@ -0,0 +1,325 @@
+      const auto& gauge_H =
+          db::get<GeneralizedHarmonic::Tags::GaugeH<Dim, Frame::Inertial>>(box);
+
+      // Compute the error in the solution, and generate tensor component list.
```
no comma

prayush replied:
this file has been removed - replaced by generic observers

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp:
```diff
@@ -0,0 +1,325 @@
+
+      // Remove tensor types, only storing individual components.
+      std::vector<TensorComponent> components;
+      components.reserve(11);  // FIXME
```
Why 11? Probably you wanted to fix that number to the actual number of components.
Is it even important to reserve the memory?

hrueter replied:
I think "Remove tensor types" is misleading here. "Combine components" might be better. 
You are also not really storing the individual components.

prayush replied:
this file has been removed - replaced by generic observers

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp:
```diff
@@ -0,0 +1,325 @@
+      using PlusSquare = funcl::Plus<funcl::Identity, funcl::Square<>>;
+
+      // FIX ME: don't copy, just initialize to the right size
+      DataVector error_in_psi_components{num_grid_points, 0.};
```
I think you can remove the FIX ME.

prayush replied:
this file has been removed - replaced by generic observers

hrueter commented on tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml:
```diff
@@ -0,0 +1,38 @@
+
+AnalyticSolution:
+  WrappedGr:
+    Mass: 1.0
```
Don't you have to specify KerrSchild as your solution here?

nilsvu replied:
Fixed in #1657

prayush replied:
fixed

hrueter reviewed (changes requested):
I did not run the input file, but the Travis tests are failing, which might be due to an error in the input file.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,195 @@
+            simple_tags,
+            db::AddComputeTags<
+                ::Tags::DerivCompute<
+                    extras_tag, ::Tags::InverseJacobian<
```
We don't see where the two simple tags (GaugeH and dtGaugeH) are updated during the evolution. That is necessary because they are simple tags and not compute tags. They are only simple tags because you need to wrap them in `Variables`.
To remove this hack using the `extras_tag` the `DerivCompute` tag should be changed such that it also accepts tensors as input. This however is something for a separate PR.
Another way of removing this hack would be to write a new compute tag "GaugeHVariablesCompute" that computes a `Variables` instead of a `Tensor`.

```
struct GaugeHVariablesCompute : db::ComputeTag, ::Tags::Variables<tmpl::list<Tags::GaugeH>>
```

prayush replied:
Hi @hrueter and @nilsleiffischer - agreed, and I have now switched to using the damped harmonic gauge; now both `GaugeH` and its spacetime derivatives are added through compute tags. I had to make some changes to the `DampedHarmonicH` code, which is committed as the first (earliest) commit here. If you like, I can make a separate PR for it (I'd prefer it stay here, but as a separate commit).

Regarding the ugly hack for computing `deriv<GaugeH>`:
 - the right solution is to modify  `DerivCompute` tag to accept tensors as input;
 - I did not want to replace it with yet another hack, and so I've removed the need for `deriv<GaugeH>` from this executable. What this means is that this exe will not observe the `3index` and `2index` constraints for now. When `DerivCompute` is properly fixed, it'd be trivial to add them back. In the meanwhile, we have all other constraints being output as well as the deviation of numerical solution from analytic solution - and these measures should be sufficient to diagnose problems / show convergence.

hrueter reviewed (changes requested):
group review by @hrueter (main) and @nilsleiffischer 

In a dynamic simulation GaugeH would be updated in every time step. I think here GaugeH is just static. It would be nice if we could use dynamic GaugeH already in this PR.
For static KerrSchild this would then be either a new gauge source function (H=H_0) or maybe this is even already contained in damped harmonic gauge.

geoffrey4444 reviewed (changes requested):
I'm happy with this, except Travis builds all fail.

geoffrey4444 commented:
Clang-tidy Travis check fails

prayush commented:
@sxs-collaboration/spectre-core-devs - could you please restart travis tests (which have failed while building spectre with "The job exceeded the maximum time limit for jobs, and has been terminated."). thanks!

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedHarmonic.cpp:
```diff
@@ -878,6 +816,96 @@ SpacetimeDerivDampedHarmonicHCompute<SpatialDim, Frame>::function(
           const double g_exponent, const int exponent) noexcept;
 
+#define INSTANTIATE_DV_FUNC(_, data)                                           \
+  template void GeneralizedHarmonic::damped_harmonic_h(                        \
```
Can you elaborate why the changes in this commit (Changes necessary to use damped harmonic gauge in an exe ) are necessary?

nilsvu replied:
ping on this

prayush replied:
The changes necessary to make damped harmonic compute tags work with exe were the split of `OptionTag`s into `Tag`s + `OptionTag`s. Others are clean ups made to the damped harmonic code: previously the compute tags were being explicitly instantiated, which were implicitly instantiating the underlying functions being called. This is not the convention in the rest of spectre though, where the functions underlying compute tags are instead instantiated. Therefore I cleaned that up here.

hrueter replied:
Please at these clarifications to the commit message.

prayush replied:
Sounds good, when I'm rebasing the new fixups I'll reword the first commit message to include these details.. Btw @nilsleiffischer , @hrueter - I have addressed all your comments with the last commit I believe, please have a look and let me know if we're ready to rebase + merge now..

nilsvu replied:
ping on this

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -222,11 +276,12 @@ namespace OptionTags {
  * the coordinate time at which roll-on begins.
  */
-struct GaugeHRollOnStartTime : db::SimpleTag {
+struct GaugeRollOnStart : Tags::GaugeHRollOnStartTime {
```
Please rebase on the newest develop to use @kidder 's new option handling.
```
namespace OptionTags{
TheOption{
type
help
}
}

namespace Tags{
TheTag{
type
name
options = tmpl::list<OptionsTags::TheOption>
static type create_from_options(const type& option_value) noexcept {
return option_value;
}
}
}
```

prayush replied:
Done

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+                    const ParallelComponent* const /*meta*/) noexcept {
+    return std::make_tuple(
+        Initialization::merge_into_databox<InitializeConstraintsTags,
+                                           simple_tags, compute_tags>(
```
Please inline the (empty) `simple_tags`.
Put the `compute_tags` list directly before the `return` statement. There is no convention for this yet. It is only our recommendation to make things more readable.

prayush replied:
changed

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+
+template <size_t Dim>
+struct InitializeGHAnd3Plus1VariablesTags {
+  using initialization_option_tags = tmpl::list<>;
```
delete

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+namespace Actions {
+template <size_t Dim>
+struct InitializeConstraintsTags {
+  using initialization_option_tags = tmpl::list<>;
```
delete

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+struct InitializeConstraintsTags {
+  using initialization_option_tags = tmpl::list<>;
+
+  using Inertial = Frame::Inertial;
```
write
```
using frame = Frame::Inertial;
```
instead

prayush replied:
changed

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+
+  using Inertial = Frame::Inertial;
+  using system = GeneralizedHarmonic::System<Dim>;
+  using variables_tag = typename system::variables_tag;
```
Also move this into the `apply` function, then you can get `system` from the `Metavariables`.

prayush replied:
done

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+
+    // call compute item for the initial gauge source function
+    const auto initial_gauge_h =
+        GeneralizedHarmonic::Tags::GaugeHImplicitFrom3p1QuantitiesCompute<
```
please call `GeneralizedHarmonic::gauge_source` directly.

prayush replied:
done

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+    // compute spatial derivatives of InitialGaugeH
+    using extras_tag = ::Tags::Variables<tmpl::list<
+        GeneralizedHarmonic::Tags::InitialGaugeH<Dim, Frame::Inertial>>>;
+    using ExtraVars = db::item_type<extras_tag>;
```
using InitialGaugeHVars =  ::Variables<tmpl::list<GeneralizedHarmonic::Tags::InitialGaugeH<Dim, Frame::Inertial>>>

Then remove `extras_tag`.
We felt the name `ExtraVars` was too unspecific.

prayush replied:
done

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,227 @@
+    const auto d_initial_gauge_source = get<
+        ::Tags::deriv<GeneralizedHarmonic::Tags::InitialGaugeH<Dim, Inertial>,
+                      tmpl::size_t<Dim>, Inertial>>(
+        partial_derivatives<typename db::item_type<extras_tag>::tags_list>(
```
use `ExtraVars` (or better the new name for that) directly

hrueter commented on tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml:
```diff
@@ -0,0 +1,50 @@
+
+# Executable: EvolveGeneralizedHarmonic
+# Check: execute
+# Timeout: 4
```
Can you use fewer grid points to stay under 2 s?

geoffrey4444 replied:
Reducing the number of grid points doesn't significantly change the run time, nor does reducing the final time. The only thing I found that reduces the run time at all is going from 1 refinement to 0 refinements, but i) that's not great, because then there are not interior boundaries in the test, and ii) it still only reduces the run time by about 1/3, whereas to get from 4s to 2s I would need a reduction by a factor of 1/2. So I think for now we should leave the timeout at 4.

hrueter replied:
Do you know, what makes this test so slow?

nilsvu replied:
ping on this

prayush replied:
What I've seen is that with debugging symbols enabled (i.e. in `Debug` mode), the run time of this test increases by a factor of 4+. My best guess is that loading these symbols initially during execution takes most of the extra time, especially due to the large number of compute tags involved.

hrueter commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,289 @@
+struct EvolutionMetavars {
+  // Customization/"input options" to simulation
+  static constexpr int volume_dim = 3;
+  using Inertial = Frame::Inertial;
```
write
```
using frame = Frame::Inertial;
```
instead

prayush replied:
done

hrueter reviewed (dismissed):
group review by @hrueter (main) and @nilsleiffischer

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -46,7 +45,7 @@ namespace GeneralizedHarmonic {
 template <size_t Dim>
 struct InitializeConstraintsTags {
-  using initialization_option_tags = tmpl::list<>;
+  using initialization_tags = tmpl::list<>;
```
delete these empty lists. They are empty by default.

prayush replied:
Yes - but I have a mild preference for keeping it explicit for aesthetic completeness and easier readibility for newcomers, unless there's actually a runtime performance  hit.

hrueter commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -276,7 +291,7 @@ struct GeneralizedHarmonicGroup {
-struct GaugeRollOnStart : Tags::GaugeHRollOnStartTime {
+struct GaugeRollOnStart {
   using type = double;
   static std::string name() noexcept { return "GaugeRollOnStart"; }
```
You don't have to specify a `name` if it is coincident with the `struct` name.

prayush replied:
Yes - but I have a mild preference for keeping it explicit for aesthetic completeness and easier readibility for newcomers, unless there's actually a runtime performance hit.

wthrowe replied:
(Not a binding request, just an opinion:) The reason I prefer implicit names is that you can't screw them up with typos or, more commonly, by forgetting to change them when the struct name changes (such as when the struct is copy-pasted as a template for a new tag struct).

nilsvu replied:
I agree with @wthrowe

prayush replied:
Okay, this is convincing enough for me, and so I've removed `name`s from OptionTags where it is redundant.

hrueter reviewed (commented):
Group review by @hrueter  (main) and @nilsleiffischer 

You can squash the current fixups.

nilsvu commented:
@prayush please ping us when you have addressed the review comments, and please don‘t squash the changes directly but post them as fixups (makes reviewing way easier)

geoffrey4444 commented:
Prayush and I are both working on this PR...one of us will ping you once change requests are addressed. Meanwhile, I just pushed a rebase that fixes the conflicts and allows the branch to compile, so I can check if travis is ok.

We'll push further changes as fixup commits

prayush commented:
Thanks @geoffrey4444 for fixing conflicts. @nilsleiffischer - I think I have pushed fixes for all remaining review comments. Please do have a look!

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -204,6 +193,13 @@ struct InitializeGaugeTags {
+    // Add all gauge tags
+    using simple_tags = db::AddSimpleTags<
+        GeneralizedHarmonic::Tags::InitialGaugeH<Dim, frame>,
+        GeneralizedHarmonic::Tags::SpacetimeDerivInitialGaugeH<Dim, frame>>;
```
please inline the `simple_tags` below (in `merge_into_databox`). That makes it easier to understand what tags the arguments to `merge_into_databox` correspond too.

prayush replied:
inlined

nilsvu reviewed (changes requested):
Please squash this fixup (including the one comment below)

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+        get<gr::Tags::TraceSpatialChristoffelFirstKind<Dim, frame, DataVector>>(
+            box);
+
+    // call compute item for the initial gauge source function
```
you can delete this comment, or update it to reflect that you're not using the compute item anymore but the free function

prayush replied:
updated

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+    const auto dt_initial_gauge_source =
+        make_with_value<tnsr::a<DataVector, Dim, frame>>(lapse, 0.);
+
+    // compute spatial derivatives of InitialGaugeH
```
please add to this comment that you need the Variables for the `partial_derivatives` function, which doesn't support single tensors (yet)

prayush replied:
added

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+
+    get<GeneralizedHarmonic::Tags::InitialGaugeH<Dim, frame>>(
+        initial_gauge_h_vars) = initial_gauge_h;
+    const auto inverse_jacobian = db::get<
```
`const auto&`

prayush replied:
changed

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+        get<::Tags::deriv<GeneralizedHarmonic::Tags::InitialGaugeH<Dim, frame>,
+                          tmpl::size_t<Dim>, frame>>(
+            partial_derivatives<typename InitialGaugeHVars::tags_list>(
+                initial_gauge_h_vars, mesh, inverse_jacobian));
```
you can `std::move` the initial_gauge_h_vars

prayush replied:
changed

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+    const auto initial_d4_gauge_h =
+        GeneralizedHarmonic::Tags::SpacetimeDerivGaugeHCompute<
+            Dim, frame>::function(dt_initial_gauge_source,
+                                  d_initial_gauge_source);
```
these two variables you can also `std::move`

prayush replied:
moved

nilsvu commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,289 @@
+          Tags::BoundaryDirectionsInterior<volume_dim>>,
+      dg::Actions::ImposeDirichletBoundaryConditions<EvolutionMetavars>,
+      dg::Actions::ReceiveDataForFluxes<EvolutionMetavars>,
+      tmpl::conditional_t<local_time_stepping, tmpl::list<>,
```
have you tested local time stepping with this executable? If not, perhaps better remove these conditionals for now?

prayush replied:
done

nilsvu commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,289 @@
+          system,
+          dg::Initialization::slice_tags_to_face<
+              typename system::variables_tag,
+              gr::Tags::SpatialMetricCompute<volume_dim, frame, DataVector>,
```
please slice the simple tags, not the compute tags (i.e. `gr::Tags::SpatialMetric` for this particular case).

prayush replied:
changed

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+namespace GeneralizedHarmonic {
+namespace Actions {
+template <size_t Dim>
+struct InitializeConstraintsTags {
```
remove the `Tags` from the name (and from the other actions in this file)

prayush replied:
removed

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,214 @@
+};
+
+template <size_t Dim>
+struct InitializeGHAnd3Plus1VariablesTags {
```
I can't say that I like this name very much but don't have a better suggestion 😄 But make it CamelCase please: `...Gh...`

prayush replied:
changed to proper camel case

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -55,6 +56,65 @@ struct ConstraintGamma2 : db::SimpleTag {
+ */
+struct GaugeHRollOnStartTime : db::SimpleTag {
+  using type = double;
+  static std::string name() noexcept { return "GaugeHRollOnStartTime"; }
```
you can delete these redundant names

prayush replied:
I have deleted `name()`s from the `OptionTags` below, but there are more than 15 `SimpleTags` in this file which have `name()` explicitly defined. If we are trying to enforce a convention here for `SimpleTags` that should be done **consistently** in a separate PR; so I prefer to keep the `name()` in here for now to conform to the file's convention. I am happy to follow this PR up with a quick un-`name`-ing PR, of course :)

nilsvu replied:
Ok I‘m fine with either. There a lot of redundant `name()`s in the code now and it‘s not particularly urgent to clean them up.

prayush replied:
Ok, then let's keep them in simple tags for now

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -55,6 +56,65 @@ struct ConstraintGamma2 : db::SimpleTag {
+struct GaugeHRollOnStartTime : db::SimpleTag {
+  using type = double;
+  static std::string name() noexcept { return "GaugeHRollOnStartTime"; }
+  using option_tags = tmpl::list<OptionTags::GaugeRollOnStart>;
```
Can you make the struct names consistent please, i.e. name them `GaugeH` or `Gauge` consistently? I think it's `GaugeH` in most of the code right now

prayush replied:
made `GaugeH` uniform throughout

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -213,6 +273,15 @@ struct ConstraintEnergy : db::SimpleTag {
+ * \ingroup OptionGroupsGroup
+ * Groups option tags related to the ValenciaDivClean evolution system.
+ */
+struct GeneralizedHarmonicGroup {
```
You can remove the `GeneralizedHarmonic` from this name since it's already in the namespace (just name it `Group`.

Also, please fix the documentation and the help string

prayush replied:
done

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -222,11 +291,11 @@ namespace OptionTags {
  * the coordinate time at which roll-on begins.
  */
-struct GaugeHRollOnStartTime : db::SimpleTag {
+struct GaugeRollOnStart {
```
When making the names consistent between tags and option tags, you can add a `name` function to shorten the names of those option tags that otherwise would have too long names.

You could also put these gauge parameters in another option-subgroup within the `GeneralizedHarmonic` group, perhaps a struct named `GaugeGroup` with the option-name `Gauge`. Then you can remove the `GaugeH` prefix from all the `name`s so you have more characters available to describe the parameter

prayush replied:
@nilsleiffischer -  Do you mean something like the following?

```
namespace OptionTags {                                                          
/*!                                                                             
 * \ingroup OptionGroupsGroup                                                   
 * Groups option tags related to the GeneralizedHarmonic evolution system.      
 */                                                                             
struct Group {               // <-- name change here                                 
  static std::string name() noexcept { return "GeneralizedHarmonic"; }          
  static constexpr OptionString help{"Options for the evolution system"};       
  using group = ::OptionTags::EvolutionSystemGroup;                             
};                                                                              
                                                                      
struct GaugeGroup {      // <-- new subgroup                                         
  static std::string name() noexcept { return "GeneralizedHarmonic"; }          
  static constexpr OptionString help{"Options for the evolution system"};       
  using group = GeneralizedHarmonic::OptionTags::Group;                             
}; 

/*!                                                                             
 * \ingroup OptionTagsGroup                                                     
 * \brief Gauge control parameter determining when to start rolling-on the      
 * evolution gauge.                                                             
 *                                                                              
 * \details The evolution gauge is gradually transitioned to (or                
 * *rolled-on* to) at the beginning of an evolution. This parameter sets        
 * the coordinate time at which roll-on begins.                                 
 */                                                                             
struct RollOnStart {               // <-- name change here                                                                                       
  using type = double;                                                          
  static constexpr OptionString help{                                           
      "Simulation time to start rolling-on evolution gauge"};                   
  using group = GaugeGroup;                                       
};                                                                              
                                                                                
struct RollOnWindow {               // <-- name change here                                                                                      
  using type = double;                                                          
  static constexpr OptionString help{                                           
      "Duration of gauge roll-on in simulation time"};                          
  using group = GaugeGroup;                                       
};                                                                              

template <typename Frame>                                                       
struct SpatialDecayWidth {                // <-- name change here                                                                                
  using type = double;                                                          
  static constexpr OptionString help{                                           
      "Spatial width of weighting factor in evolution gauge"};                  
  using group = GaugeGroup;                                       
};          
```

nilsvu replied:
Yep, just keep the `Gauge` in the struct names:

```cpp
struct GaugeRollOnStart {
    static std::string name() noexcept { return „RollOnStart“; }
    // ...
}
```

prayush replied:
Okay, cool. Just to make sure I'm reading this right, now there's Gauge in struct names but not in their `name()`. Which means the `name` method gets added back everywhere. Yes?

prayush replied:
Also, do you want "Gauge" in their names or "GaugeH"?

nilsvu replied:
- Yes, gauge in struct names (since they are not in an extra namespace and not length-constrained), and yes you need a name method for the option tags in the gauge group
- Just make Gauge vs GaugeH consistent between tag names :)

prayush replied:
done.

nilsvu commented on tests/Unit/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Test_DampedHarmonicGaugeQuantities.cpp:
```diff
@@ -2955,21 +2960,29 @@ void test_damped_harmonic_compute_tags(const size_t grid_size_each_dimension,
+  CHECK(db::get<GeneralizedHarmonic::Tags::GaugeHRollOnTimeWindow>(box) ==
+        sigma_t_S);
+  CHECK(db::get<GeneralizedHarmonic::Tags::GaugeHSpatialWeightDecayWidth<
+            Frame::Inertial>>(box) == r_max);
```
@wthrowe has recently removed these checks because we can trust the DataBox to store the simple tags added a few lines above (this functionality is tested elsewhere). It probably got re-added during the rebase, so please remove it.

prayush replied:
removed them

nilsvu commented on cmake/AddInputFileTests.cmake:
```diff
@@ -98,12 +98,24 @@ function(add_input_file_tests INPUT_FILE_DIR)
 
+    # Read the timeout duration specified in input file, empty is accepted.
+    # The default duration is 2 seconds.
+    string(REGEX MATCH "#[ ]*Timeout:[^\n]+"
```
Please make a quick separate PR that adds this timeout to input files to make this change visible to the other devs. (We discussed changes that should be stand-alone PRs recently in #1689)

geoffrey4444 replied:
#1695

prayush replied:
#1695 got merged, and I've rebased this PR onto it.

nilsvu commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,295 @@
+#include "Domain/Tags.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "ErrorHandling/FloatingPointExceptions.hpp"
+#include "Evolution/Actions/ComputeTimeDerivative.hpp"  // IWYU pragma: keep
```
Have you run IWYU over these files? If no, please remove the pragmas. If yes, good :)

geoffrey4444 replied:
No, so we'll remove the pragmas

nilsvu commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,295 @@
+      // When using local timestepping, replace the previous tag with the
+      // following two lines.
+      // Tags::TimeStepper<tmpl::conditional_t<local_time_stepping,
+      //                                       LtsTimeStepper, TimeStepper>>,
```
I asked this in another comment before, have you actually tested local time stepping with this executable? If no, please just remove this code.

geoffrey4444 replied:
No, we haven't tested with local time stepping

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Tags.hpp:
```diff
@@ -222,11 +302,12 @@ namespace OptionTags {
  * the coordinate time at which roll-on begins.
  */
-struct GaugeHRollOnStartTime : db::SimpleTag {
+struct GaugeHRollOnStart {
```
One last comment on these tag names: I think you originally changed these names to make them shorter to fit in options. You don’t need to do that anymore since you implement a name function.

geoffrey4444 replied:
I prefer just leaving the names as they are right now

prayush replied:
I made this name shorter so as to not end up with the same name for the `SimpleTag` and the `OptionTag`, which are `GaugeHRollOnStartTime` and `GaugeHRollOnStart` right now. As the `name()` of the option tag has been cleaned up anyway, lets leave them as they are. We can change them later if you like, but this shouldn't stop this PR from being merged.

nilsvu replied:
Ok, I won't insist on changing these names

nilsvu reviewed (changes requested):
Please squash the current fixups.

prayush committed:
Changes to use damped harmonic gauge in an exe (+ cleanups)

One change made here necessary to make damped harmonic compute
tags work with the GH exe, i.e. the split of its OptionTags into
SimpleTags + OptionTags.

Others changes in here are clean ups made to the damped harmonic
code: (A) clean up documentation / naming of the option tags;
(B) previously the damped harmonic compute tags were being
explicitly instantiated, which were implicitly instantiating
the underlying functions being called. As this is not the
convention in the rest of spectre, where functions underlying
compute tags are instead instantiated, that was cleaned up here.

prayush committed:
Add GH executable for analytic solution initial data

prayush commented:
thanks @geoffrey4444; 

@nilsleiffischer @hrueter - assuming the review is now near convergence, I have rearranged code changes in two clean commits and rebased recent changes appropriately.

nilsvu reviewed (commented):
Yep, this has come a long way and looks pretty much ready now. Please remember not to immediately squash requested changes, though (most recently Geoffrey's fixup was squashed in before I had a chance to look at it). I understand you want to finally get this merged, and pushing to this PR from two different branches complicates things, but for reviewers it's hard to keep an overview when they can't be sure what code has actually changed.

The one comment below is about the commit message that you wanted to add details to.

nilsvu commented:
@sxs-collaboration/spectre-core-devs Recently only Hannes and myself have reviewed this PR. Could you provide a second review and get this merged?

prayush commented:
> Yep, this has come a long way and looks pretty much ready now. Please remember not to immediately squash requested changes, though (most recently Geoffrey's fixup was squashed in before I had a chance to look at it). I understand you want to finally get this merged, and pushing to this PR from two different branches complicates things, but for reviewers it's hard to keep an overview when they can't be sure what code has actually changed.
> 
oops yes, it was pushing from two branches that got it.

> The one comment below is about the commit message that you wanted to add details to.

I've padded the commit message with a suitable explanation.

osheamonn commented:
I will be able to do my requested review either today or tomorrow. :+1:

geoffrey4444 commented:
Travis fails because ObserveErrorNorms.hpp not found, not sure why

prayush commented:
> Travis fails because ObserveErrorNorms.hpp not found, not sure why

pushed a fix!

osheamonn reviewed (dismissed):
This all looks good. :+1:

prayush commented:
Hi @sxs-collaboration/spectre-core-devs - this PR has requisite review approvals, passes all travis tests, and is ready to merge. Can we please get this into develop?

nilsvu commented:
@prayush please squash your last commit 20c646e into the commit where you add executable (note to other reviewers, it increases the timeout for the input file test to 5 seconds)

prayush commented:
> @prayush please squash your last commit [20c646e](https://github.com/sxs-collaboration/spectre/commit/20c646e83833ca79dd52b8b87554da5c86d24a30) into the commit where you add executable (note to other reviewers, it increases the timeout for the input file test to 5 seconds)

Done. 

Note to reviewers: this PR had approvals from at least 2 reviewers before this squash (which has now dismissed them).

kidder commented:
@nilsleiffischer @hrueter please look at

nilsvu commented on tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml:
```diff
@@ -0,0 +1,54 @@
+
+Observers:
+  VolumeFileName: "VolumeData"
+  ReductionFileName: "ReductionData"
```
When you get the chance, in a follow-up PR, please prefix these file names with `GeneralizedHarmonic` to avoid race conditions.

nilsvu commented:
@kidder, I think we don't need to wait for approval by @hrueter since he is at a workshop for a few weeks and may not have time to look at this (@hrueter if you do have some time in the next day or two, please take a look). Recently we reviewed this together, anyway.

If we get this in before #1703 then we can avoid the rebase onto that (and I do it in #1703 instead).

prayush commented:
@sxs-collaboration/spectre-core-devs - this is now ready to merge. Could you please have a look?

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedHarmonic.cpp:
```diff
@@ -877,6 +815,97 @@ SpacetimeDerivDampedHarmonicHCompute<SpatialDim, Frame>::function(
+  template typename db::item_type<                                             \
+      GeneralizedHarmonic::Tags::GaugeH<DIM(data), FRAME(data)>>               \
+  GeneralizedHarmonic::DampedHarmonicHCompute<DIM(data), FRAME(data)>::        \
+      function(const typename db::item_type<                                   \
```
You don't need to explicitly instantiate functions defined in the hpp

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedHarmonic.hpp:
```diff
@@ -286,4 +258,64 @@ void spacetime_deriv_damped_harmonic_h(
+      Tags::GaugeHSpatialWeightDecayWidth<Frame>>;
+
+  static constexpr
+      typename db::item_type<Tags::SpacetimeDerivGaugeH<SpatialDim, Frame>>
```
you don't need `typename` in front of `db::item_type` because it's a type alias

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,276 @@
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/WaveEquation/PlaneWave.hpp"
```
Probably don't need this include

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,276 @@
+      GeneralizedHarmonic::Actions::InitializeConstraints<volume_dim>,
+      dg::Actions::InitializeMortars<EvolutionMetavars, true>,
+      Initialization::Actions::DiscontinuousGalerkin<EvolutionMetavars>,
+      Initialization::Actions::Minmod<volume_dim>,
```
I don't know for sure, but you probably aren't using a limiter

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,276 @@
+static const std::vector<void (*)()> charm_init_node_funcs{
+    &setup_error_handling,
+    &domain::creators::register_derived_with_charm,
+    &Parallel::register_derived_classes_with_charm<MathFunction<1>>,
```
Is this left over from the executable this was based on? You might be able to get rid of a corresponding include as well.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+#include "ErrorHandling/Assert.hpp"
+#include "Evolution/Initialization/DiscontinuousGalerkin.hpp"
+#include "Evolution/Initialization/Evolution.hpp"
+#include "Evolution/Initialization/Limiter.hpp"
```
You can probably remove the limiter

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+#include "Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedHarmonic.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/System.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/Divergence.tpp"
```
You probably don't need `Divergence`, but should use the `hpp` if you do need it, then do an explicit instantiation in a `cpp` file.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    // compute initial-gauge related quantities
+    const auto mesh = db::get<::Tags::Mesh<Dim>>(box);
```
`const auto&`

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+        spatial_metric, trace_extrinsic_curvature,
+        trace_christoffel_last_indices);
+    // set time derivatives of InitialGaugeH = 0
+    // NOTE: this will need to be generalized to handle numerical initial data
```
Could you file an issue about this so we have a record outside of the file?

prayush replied:
#1708

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+    const auto& inverse_jacobian = db::get<
+        ::Tags::InverseJacobian<::Tags::ElementMap<Dim, frame>,
+                                ::Tags::Coordinates<Dim, Frame::Logical>>>(box);
+    const auto d_initial_gauge_source =
```
non-const because you move it later

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+                std::move(initial_gauge_h_vars), mesh, inverse_jacobian));
+
+    // compute spacetime derivatives of InitialGaugeH
+    const auto initial_d4_gauge_h =
```
non-const because you move it later

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+        get<::Tags::deriv<GeneralizedHarmonic::Tags::InitialGaugeH<Dim, frame>,
+                          tmpl::size_t<Dim>, frame>>(
+            partial_derivatives<typename InitialGaugeHVars::tags_list>(
+                std::move(initial_gauge_h_vars), mesh, inverse_jacobian));
```
the move here doesn't help us, unfortunately.

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+            box);
+
+    // compute the initial gauge source function
+    const auto initial_gauge_h = GeneralizedHarmonic::gauge_source<Dim, frame>(
```
non-const since you move it later

nilsdeppe commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -0,0 +1,217 @@
+    // set time derivatives of InitialGaugeH = 0
+    // NOTE: this will need to be generalized to handle numerical initial data
+    // and analytic initial data whose gauge is not initially stationary.
+    const auto dt_initial_gauge_source =
```
non-const since you move this later

nilsdeppe reviewed (commented):
Review with @kidder and @wthrowe 

We're happy for you to squash immediately

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -0,0 +1,272 @@
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Ricci.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "PointwiseFunctions/MathFunctions/MathFunction.hpp"
```
This is a ping on the "corresponding" include part of https://github.com/sxs-collaboration/spectre/pull/1500#discussion_r330766985 Can you get rid of the MathFunctions include?

geoffrey4444 replied:
Can we please do this in a quick follow up PR? I’ll open an issue and assign myself, and I’ll do it promptly after 1500 gets merged.

geoffrey4444 replied:
Never mind...I'll just do it

prayush replied:
Hi @geoffrey4444 @nilsdeppe - removed the include :)

geoffrey4444 replied:
Looks like Prayush beat me to it...I was double-checking that it would compile

nilsdeppe replied:
:+1:

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+    IO
+    Informer
+    LinearOperators
+    MathFunctions
```
We can figure out whether or not this linker dependency to MathFunctions is needed later :)

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/DampedHarmonic.cpp:
```diff
@@ -321,11 +321,9 @@ spacetime_deriv_of_power_log_factor_metric_lapse(
         gauge_h,
-    const typename db::item_type<Tags::InitialGaugeH<SpatialDim, Frame>>&
-        gauge_h_init,
+    const db::item_type<Tags::InitialGaugeH<SpatialDim, Frame>>& gauge_h_init,
```
Write the tensor type explicitly where you know it, instead of resorting to item_type. Helps compile times (I think) and also decouples functions that don‘t need to know anything about databoxes and tags from those that do.

nilsvu commented on src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp:
```diff
@@ -496,16 +497,58 @@ void UpwindFlux<Dim>::operator()(
+
+template <size_t Dim>
+using variables_tags =
+    typename GeneralizedHarmonic::System<Dim>::variables_tag::tags_list;
```
Use `db::get_variables_tags_list`, which can also handle prefixed variables tags. Not strictly necessary here, but may help someone who copies code around avoid a compiler error :)

nilsvu reviewed (dismissed):
Just a few more notes that you can address in follow-up PRs

nilsvu commented:
@prayush to fix the compiler error on Travis, use `evolution::OptionTags::SystemGroup`

prayush commented:
It seems that we're ready to merge this @sxs-collaboration/spectre-core-devs !

geoffrey4444 commented:
Thanks @prayush for all of your hard work getting this PR ready to merge, and thanks everyone for your helpful review comments!
